Windows build fails on Flutter 3.41.6 / VS 2022 in audio_decoder_plugin.cpp (C4245 treated as error, plus C2589/C2059)

## Describe the bug

Building a Flutter app for Windows fails in the `audio_decoder` Windows plugin.

The failure happens in the generated plugin symlink path:

`windows/flutter/ephemeral/.plugin_symlinks/audio_decoder/windows/audio_decoder_plugin.cpp`

This appears to be in the plugin's Windows C++ code rather than in my Dart app code.

## Package version

`audio_decoder: 0.7.4`

## Build output

```text
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(780,9): error C2220: the following warning is treated as an error [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(780,9): warning C4245: 'argument': conversion from '__MIDL___MIDL_itf_mfreadwrite_0000_0001_0001' to 'DWORD', signed/unsigned mismatch [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(905,22): error C2589: '(': illegal token on right side of '::' [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(905,83): error C2059: syntax error: ')' [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(953,36): error C2589: '(': illegal token on right side of '::' [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(953,36): error C2059: syntax error: ')' [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(960,27): error C2589: '(': illegal token on right side of '::' [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
P:\Projects\LearnLocalMusic\windows\flutter\ephemeral\.plugin_symlinks\audio_decoder\windows\audio_decoder_plugin.cpp(960,27): error C2059: syntax error: ')' [P:\Projects\LearnLocalMusic\build\windows\x64\plugins\audio_decoder\audio_decoder_plugin.vcxproj]
Error: Build process failed.

--- a/linux/audio_decoder_plugin.cc
+++ b/linux/audio_decoder_plugin.cc
@@ -583,9 +583,16 @@ static FlValue* GetWaveform(const std::string& path, int numberOfSamples) {
     (G_TYPE_CHECK_INSTANCE_CAST((obj), audio_decoder_plugin_get_type(), \
                                 AudioDecoderPlugin))
 
+typedef struct _AudioDecoderPlugin AudioDecoderPlugin;
+typedef struct _AudioDecoderPluginClass AudioDecoderPluginClass;
+
 struct _AudioDecoderPlugin {
     GObject parent_instance;
     FlMethodChannel* channel;
+};
+
+struct _AudioDecoderPluginClass {
+    GObjectClass parent_class;
 };
 
 G_DEFINE_TYPE(AudioDecoderPlugin, audio_decoder_plugin, g_object_get_type())

--- a/windows/audio_decoder_plugin.cpp
+++ b/windows/audio_decoder_plugin.cpp
@@ -19,6 +19,7 @@
 #include <functional>
 #include <cmath>
 #include <algorithm>
+#include <cctype> 
 
 #pragma comment(lib, "mfplat.lib")
 #pragma comment(lib, "mfreadwrite.lib")
@@ -847,7 +848,7 @@ std::string AudioDecoderPlugin::TrimAudio(
     std::string ext = outputPath.substr(outputPath.find_last_of('.') + 1);
     std::transform(ext.begin(), ext.end(), ext.begin(),
 		    [](unsigned char c) {
-		    	return static_cast<char>(std::tolower());
+		    	return static_cast<char>(std::tolower(c));
 		});
 
     if (ext == "m4a") {

--- a/windows/audio_decoder_plugin.cpp
+++ b/windows/audio_decoder_plugin.cpp
@@ -777,7 +777,7 @@ flutter::EncodableMap AudioDecoderPlugin::GetAudioInfo(const std::string& path) 
 
     IMFMediaSource* pSource = nullptr;
     hr = pReader->GetServiceForStream(
-        MF_SOURCE_READER_MEDIASOURCE, GUID_NULL,
+        static_cast<DWORD>(MF_SOURCE_READER_MEDIASOURCE), GUID_NULL,
         __uuidof(IMFMediaSource), reinterpret_cast<LPVOID*>(&pSource));
     if (SUCCEEDED(hr) && pSource) {
         IMFPresentationDescriptor* pPD = nullptr;
@@ -902,7 +902,7 @@ std::string AudioDecoderPlugin::TrimAudio(
 
         for (size_t offset = 0; offset < pcm.data.size(); offset += chunkSize) {
             DWORD thisChunk = static_cast<DWORD>(
-                std::min(static_cast<size_t>(chunkSize), pcm.data.size() - offset));
+                (std::min)(static_cast<size_t>(chunkSize), pcm.data.size() - offset));
 
             IMFMediaBuffer* pBuffer = nullptr;
             MFCreateMemoryBuffer(thisChunk, &pBuffer);
@@ -950,14 +950,14 @@ flutter::EncodableList AudioDecoderPlugin::GetWaveform(
 
     const int16_t* samples = reinterpret_cast<const int16_t*>(pcm.data.data());
     size_t totalSamples = pcm.data.size() / 2;
-    size_t samplesPerWindow = std::max(static_cast<size_t>(1), totalSamples / numberOfSamples);
+    size_t samplesPerWindow = (std::max)(static_cast<size_t>(1), totalSamples / numberOfSamples);
 
     std::vector<double> waveform;
     double maxRms = 0;
 
     for (int i = 0; i < numberOfSamples; i++) {
         size_t start = static_cast<size_t>(i) * totalSamples / numberOfSamples;
-        size_t end = std::min(start + samplesPerWindow, totalSamples);
+        size_t end = (std::min)(start + samplesPerWindow, totalSamples);
         if (start >= totalSamples) break;
 
         double sumSquares = 0;

--- a/windows/audio_decoder_plugin.cpp
+++ b/windows/audio_decoder_plugin.cpp
@@ -845,7 +845,10 @@ std::string AudioDecoderPlugin::TrimAudio(
     }
 
     std::string ext = outputPath.substr(outputPath.find_last_of('.') + 1);
-    std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
+    std::transform(ext.begin(), ext.end(), ext.begin(),
+		    [](unsigned char c) {
+		    	return static_cast<char>(std::tolower());
+		});
 
     if (ext == "m4a") {
         // M4A trimming uses DecodeToPcm (full buffering) because

--- a/windows/audio_decoder_plugin.h
+++ b/windows/audio_decoder_plugin.h
@@ -3,6 +3,7 @@
 
 #include <flutter/method_channel.h>
 #include <flutter/plugin_registrar_windows.h>
+#include <flutter/encodable_value.h>
 
 #include <functional>
 #include <memory>

